### ISSUE TITLE
count can't interpolate data sources in some cases

### ISSUE THREAD

**Issue by brikis98** [hashicorp/terraform#17812]
### Terraform Version

Terraform v0.11.3

### Terraform Configuration Files

I have a module called `load-balancer` that creates a load balancer and some security group rules. The code that's breaking has to do with the security group rules:

```hcl
resource "aws_security_group_rule" "allow_https_inbound_from_security_groups" {
  count                    = "${length(var.https_listener_ports_and_certs) * length(var.allow_inbound_from_security_groups)}"
  type                     = "ingress"
  from_port                = "${lookup(var.https_listener_ports_and_certs[count.index], "port")}"
  to_port                  = "${lookup(var.https_listener_ports_and_certs[count.index], "port")}"
  protocol                 = "tcp"
  security_group_id        = "${aws_security_group.sg.id}"
  source_security_group_id = "${element(var.allow_inbound_from_security_groups), count.index)}"
}

variable "https_listener_ports_and_certs" {
  type = "list"
}

variable "allow_inbound_from_security_groups" {
  type = "list"
}
```

Here's how I'm using the module:

```hcl
module "load_balancer" {
  source "./load-balancer"

  https_listener_ports_and_certs = [
    {
      port            = 4444
      certificate_arn = "${data.aws_acm_certificate.load_balancer.arn}"
    },
    {
      port            = 5555
      certificate_arn = "${data.aws_acm_certificate.load_balancer.arn}"
    }
  ]

  allow_inbound_from_security_groups = []
}
```

The key thing to note is that: 

1. `https_listener_ports_and_certs` is a map that references some data sources and variables, both items that I thought were valid to use in `count` parameter interpolations.
1. `allow_inbound_from_security_groups` is empty.

### Expected Behavior

`terraform apply` works and creates my load balancer.

### Actual Behavior

I get this error:

```
Error: module.load_balancer.aws_security_group_rule.allow_https_inbound_from_security_groups: 1 error(s) occurred:

* module.load_balancer.aws_security_group_rule.allow_https_inbound_from_security_groups: element: element() may not be used with an empty list in:

${element(var.allow_inbound_from_security_groups, count.index)}
```

It's true that `allow_inbound_from_security_groups` is empty, but that `element()` should never have been processed in the first place, as `allow_inbound_from_security_groups` is empty, and therefore, we should've gotten `count = 0` for `allow_https_inbound_from_security_groups`.

### Steps to Reproduce

1. `terraform init`
2. `terraform apply`

### Additional Context

If I remove the data sources from `https_listener_ports_and_certs`, or if I hardcode the `count` parameter for `allow_https_inbound_from_security_groups` to 0, then `apply` works fine.

So it seems like the issue is that while `count` can interpolate data sources, it doesn't seem to be able to interpolate a map that contains data sources.

**Comment by jbardin** [comment]
Thanks @brikis98!

It turns out this is happening during validation, where the `https_listener_ports_and_certs` value is not yet known, but the `allow_inbound_from_security_groups` value is static. Therefor the validation walk over the graph will show that the `count` for `allow_https_inbound_from_security_groups` is unknown (there is no optimization yet to compute a static portion of the expression, determine it's zero, and know that only remaining operation is `*` hence the result will be 0). Since `count` is not known, validation on the `source_security_group_id` field is attempted, which has known values which happen to be invalid. 

For reference, here's a reduced test case that exhibits this issue

```
# main.tf
module "mod" {
  source = "./mod"
  list = ["${data.template_file.data.rendered}"]
  empty_list = []
}

data "template_file" "data" {
  template = "output"
}
```

```
#./mod/main.tf
resource "null_resource" "a" {
  count = "${length(var.list) * length(var.empty_list)}"
  triggers = {
    "a" = "${element(var.empty_list, count.index)}"
  }
}

variable "list" {
  type = "list"
}

variable "empty_list" {
  type = "list"
}
```

One workaround for this would be to use the pattern we've used in outputs to ensure that they are always valid, something like (possibly modified if `count.index` is known to extend beyond the list)

    "${element(concat(var.empty_list, list("")), count.index)}"


Another option could be to make sure the empty list isn't a static value:

```
empty_list = "${null_resource.none.*.id}"

resource "null_resource" "none" {
  count = 0
}
```

**Comment by brikis98** [comment]
@jbardin Thx for the explanation. 

The `"${element(concat(var. allow_inbound_from_security_groups, list("")), count.index)}"` workaround doesn't work here, as it adds an empty element to the list, which causes issues when `allow_inbound_from_security_groups` is nonempty and we actually need to loop over its contents.

The workaround I had to use was quite a bit uglier:

```hcl
source_security_group_id = "${element(split(",", length(var.allow_inbound_from_security_groups) == 0 ? "fake-id-for-workaround" : join(",", var.allow_inbound_from_security_groups)), count.index)}"
```

The basic idea is to return a fake list of length 1 if `allow_inbound_from_security_groups` is empty and to use `allow_inbound_from_security_groups` as-is when it's not. Since Terraform doesn't support using lists with ternary syntax, I also had to `join` and `split` the list to make the types work out.

**Comment by apparentlymart** [comment]
Hi again, @brikis98!

As you've probably seen by now, we've merged the configuration language improvements for v0.12.0 into master now, and so the behavior here has changed. As James mentioned, the problem here was that during validation Terraform would still check that expression (because validation is done _regardless_ of count) but `count.index` would incorrectly have the value zero.

The change we've made to this for the next release is that during validation `count.index` is now an _unknown number_, rather than a specific value. This allows Terraform to still perform type checking (it knows it's a number) but it will not raise any errors that result from the _value_ of that number, instead saving those for the plan phase. During the plan phase `count = 0` and so no further checks are performed.

I verified this behavior with the v0.12.0-alpha1 build, using a simplified configuration:

```hcl
variable "empty" {
  type    = list(string)
  default = []
}

resource "null_resource" "example" {
  count = length(var.empty)

  triggers = {
    item = var.empty[count.index]
  }
}
```

When the list is empty, `terraform apply` takes no action at all but completes successfully as expected:

```hcl
$ terraform apply

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
```

Putting some elements in that list causes resources to be created, as before:

```
$ terraform apply -var='empty=["foo"]'

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # null_resource.example[0] will be created
  + resource "null_resource" "example" {
      + id       = (known after apply)
      + triggers = {
          + "item" = "foo"
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.

...
```

These changes will be included in the forthcoming v0.12.0 final release, so I'm going to close this out. Thanks for reporting this, and for your patience while we got to the bottom of it.
